dbms/src/Flash/Mpp/AsyncRequestHandler.h

```diff
@@ -0,0 +1,398 @@
+        // TODO add random fail point to mock the fail
+        start();
+    }
+
+    // execute will be called by RPC framework so it should be as light as possible.
+    // Do not do anything after processXXX functions.
+    void execute(bool & ok) override
+    {
+        try
+        {
+            switch (stage)
+            {
+            case AsyncRequestStage::WAIT_RETRY:
+                start();
+                break;
+            case AsyncRequestStage::WAIT_MAKE_READER:
+                processWaitMakeReader(ok);
+                break;
+            case AsyncRequestStage::WAIT_BATCH_READ:
+                processWaitBatchRead(ok);
+                break;
+            case AsyncRequestStage::WAIT_REWRITE:
+                processWaitReWrite();
+                break;
+            case AsyncRequestStage::WAIT_FINISH:
+                processWaitFinish();
+                break;
+            default:
+                RUNTIME_ASSERT(false, "Unexpected stage {}", magic_enum::enum_name(stage));
+            }
+        }
+        catch (...)
+        {
+            LOG_ERROR(log, "Some errors happen in the AsyncRequestHandler.execute()");
+            closeConnection("Exception is thrown in AsyncRequestHandler");
+        }
+    }
+
+private:
+    void processWaitMakeReader(bool ok)
+    {
+        // Use lock to ensure reader is created already in reactor thread
+        std::lock_guard lock(make_reader_mu);
+        if (!ok)
+        {
+            reader.reset();
+            LOG_WARNING(log, "MakeReader fail. retry time: {}", retry_times);
+            retryOrDone("Exchange receiver meet error : send async stream request fail");
+        }
+        else
+        {
+            startAsyncRead();
+        }
+    }
+
+    void processWaitBatchRead(bool ok)
+    {
+        if (ok)
+            ++read_packet_index;
+
+        if (!ok || read_packet_index == batch_packet_count || packets[read_packet_index - 1]->hasError())
+            processReceivedData();
+        else
+            reader->read(packets[read_packet_index], thisAsUnaryCallback());
+    }
+
+    void processWaitFinish()
+    {
+        if (likely(finish_status.ok()))
+            closeConnection("");
+        else
+        {
+            // As AsyncRequestHandler may have been destructed after close_conn is called;
+            // we need to copy some data for LOG_WARNING after a while.
+            LoggerPtr copy_log = log;
+            Status copy_finish_status = finish_status;
+            int copy_retry_times = retry_times;
+
+            if (!retryOrDone(fmt::format("Exchange receiver meet error : {}", finish_status.error_message())))
+            {
+                LOG_WARNING(
+                    copy_log,
+                    "Finish fail. err code: {}, err msg: {}, retry time {}",
+                    copy_finish_status.error_code(),
+                    copy_finish_status.error_message(),
+                    copy_retry_times);
```

Comment:
It seems that if we move to https://github.com/pingcap/tiflash/pull/6811/files#diff-8f3f21fe5c9b41fa0b0ea3e1e802d3aeeca41d4a83bb01217fc9e745d7fb69aaR296, there will be no need to copy.

Reply:
And then `retryOrDone` doesn't need to return a bool.

Reply:
> It seems that if we move to https://github.com/pingcap/tiflash/pull/6811/files#diff-8f3f21fe5c9b41fa0b0ea3e1e802d3aeeca41d4a83bb01217fc9e745d7fb69aaR296, there will be no need to copy.

okk, I will move it.